File: android/src/main/kotlin/com/makemind/appplayer_core/AppPlayerCorePlugin.kt
package com.makemind.appplayer_core

import android.Manifest
import android.app.Activity
import android.content.Context
import android.content.Intent
import android.content.pm.PackageManager
import android.os.Build
import androidx.core.app.NotificationManagerCompat
import androidx.core.content.ContextCompat
import androidx.work.OneTimeWorkRequestBuilder
import androidx.work.WorkManager
import androidx.work.workDataOf
import io.flutter.embedding.engine.plugins.FlutterPlugin
import io.flutter.embedding.engine.plugins.activity.ActivityAware
import io.flutter.embedding.engine.plugins.activity.ActivityPluginBinding
import io.flutter.plugin.common.EventChannel
import io.flutter.plugin.common.MethodCall
import io.flutter.plugin.common.MethodChannel
import io.flutter.plugin.common.MethodChannel.MethodCallHandler
import io.flutter.plugin.common.MethodChannel.Result
import io.flutter.plugin.common.PluginRegistry

/**
 * Android side of the appplayer_core Platform Integration Foundation
 * (FR-PLATFORM): background execution (foreground service + WorkManager),
 * OS runtime permissions, and notifications.
 */
class AppPlayerCorePlugin :
    FlutterPlugin,
    MethodCallHandler,
    ActivityAware,
    PluginRegistry.RequestPermissionsResultListener,
    PluginRegistry.NewIntentListener {

    private lateinit var context: Context
    private lateinit var methodChannel: MethodChannel
    private lateinit var wakesChannel: EventChannel
    private lateinit var permissionChangesChannel: EventChannel
    private lateinit var notificationTapsChannel: EventChannel

    private var activity: Activity? = null
    private var activityBinding: ActivityPluginBinding? = null

    private var wakesSink: EventChannel.EventSink? = null
    private var notificationTapsSink: EventChannel.EventSink? = null

    // A tap that arrived before Dart subscribed to the taps stream (cold start);
    // flushed when the sink attaches.
    private var pendingTapSource: String? = null

    // Pending permission requests keyed by request code.
    private val pendingPermissions = mutableMapOf<Int, Result>()
    private var nextRequestCode = 1000

    private val notifications = NotificationBridge()

    override fun onAttachedToEngine(binding: FlutterPlugin.FlutterPluginBinding) {
        context = binding.applicationContext
        methodChannel = MethodChannel(binding.binaryMessenger, CHANNEL_METHODS)
        methodChannel.setMethodCallHandler(this)

        wakesChannel = EventChannel(binding.binaryMessenger, CHANNEL_WAKES)
        wakesChannel.setStreamHandler(streamHandler { wakesSink = it })

        permissionChangesChannel =
            EventChannel(binding.binaryMessenger, CHANNEL_PERMISSION_CHANGES)
        permissionChangesChannel.setStreamHandler(streamHandler { })

        notificationTapsChannel =
            EventChannel(binding.binaryMessenger, CHANNEL_NOTIFICATION_TAPS)
        notificationTapsChannel.setStreamHandler(streamHandler {
            notificationTapsSink = it
            // Flush a tap captured before the stream was listened to.
            if (it != null) {
                pendingTapSource?.let { source -> it.success(source) }
                pendingTapSource = null
            }
        })

        notifications.attach(context)
    }

    override fun onDetachedFromEngine(binding: FlutterPlugin.FlutterPluginBinding) {
        methodChannel.setMethodCallHandler(null)
        wakesChannel.setStreamHandler(null)
        permissionChangesChannel.setStreamHandler(null)
        notificationTapsChannel.setStreamHandler(null)
    }

    override fun onMethodCall(call: MethodCall, result: Result) {
        when (call.method) {
            "background.initialize" -> result.success(null)
            "background.begin" -> result.success(startBackground())
            "background.end" -> {
                stopBackground()
                result.success(null)
            }
            "background.runJob" -> {
                val jobId = call.argument<String>("jobId") ?: ""
                enqueueJob(jobId)
                result.success(null)
            }
            "permission.status" -> result.success(
                statusOf(mapPermission(call.argument<String>("permission")))
            )
            "permission.request" -> requestPermission(
                call.argument<String>("permission"), result
            )
            "notification.permissionStatus" ->
                result.success(notificationPermissionStatus())
            "notification.requestPermission" ->
                requestPermission("notifications", result)
            "notification.post" -> {
                notifications.post(
                    call.argument<String>("id") ?: "",
                    call.argument<String>("title") ?: "",
                    call.argument<String>("body") ?: "",
                    call.argument<String>("source") ?: ""
                )
                result.success(null)
            }
            "notification.cancel" -> {
                notifications.cancel(call.argument<String>("id") ?: "")
                result.success(null)
            }
            else -> result.notImplemented()
        }
    }

    // --- Background execution ------------------------------------------------

    private fun startBackground(): Boolean {
        return try {
            val intent = Intent(context, BackgroundService::class.java)
            ContextCompat.startForegroundService(context, intent)
            true
        } catch (e: Exception) {
            false
        }
    }

    private fun stopBackground() {
        context.stopService(Intent(context, BackgroundService::class.java))
    }

    private fun enqueueJob(jobId: String) {
        val request = OneTimeWorkRequestBuilder<BackgroundJobWorker>()
            .setInputData(workDataOf(BackgroundJobWorker.KEY_JOB_ID to jobId))
            .build()
        WorkManager.getInstance(context).enqueue(request)
    }

    // --- Permissions ---------------------------------------------------------

    private fun mapPermission(name: String?): String? = when (name) {
        "bluetooth" ->
            if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.S)
                Manifest.permission.BLUETOOTH_CONNECT else null
        "location" -> Manifest.permission.ACCESS_FINE_LOCATION
        "notifications" ->
            if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.TIRAMISU)
                Manifest.permission.POST_NOTIFICATIONS else null
        "camera" -> Manifest.permission.CAMERA
        "microphone" -> Manifest.permission.RECORD_AUDIO
        // backgroundExecution (foreground service), localNetwork, usb have no
        // Android runtime permission — treated as granted.
        else -> null
    }

    private fun statusOf(permission: String?): String {
        if (permission == null) return "granted"
        val granted = ContextCompat.checkSelfPermission(context, permission) ==
            PackageManager.PERMISSION_GRANTED
        return if (granted) "granted" else "denied"
    }

    private fun notificationPermissionStatus(): String {
        return if (NotificationManagerCompat.from(context).areNotificationsEnabled())
            "granted" else "denied"
    }

    private fun requestPermission(name: String?, result: Result) {
        val permission = mapPermission(name)
        if (permission == null) {
            result.success("granted")
            return
        }
        val current = activity
        if (current == null) {
            result.success(statusOf(permission))
            return
        }
        val code = nextRequestCode++
        pendingPermissions[code] = result
        current.requestPermissions(arrayOf(permission), code)
    }

    override fun onRequestPermissionsResult(
        requestCode: Int,
        permissions: Array<out String>,
        grantResults: IntArray
    ): Boolean {
        val result = pendingPermissions.remove(requestCode) ?: return false
        val granted = grantResults.isNotEmpty() &&
            grantResults[0] == PackageManager.PERMISSION_GRANTED
        result.success(if (granted) "granted" else "denied")
        return true
    }

    // --- ActivityAware -------------------------------------------------------

    override fun onAttachedToActivity(binding: ActivityPluginBinding) {
        activity = binding.activity
        activityBinding = binding
        binding.addRequestPermissionsResultListener(this)
        binding.addOnNewIntentListener(this)
        // Cold start via a notification tap: the launch intent carries the
        // source extra.
        handleTapIntent(binding.activity.intent)
    }

    override fun onNewIntent(intent: Intent): Boolean {
        handleTapIntent(intent)
        return false
    }

    private fun handleTapIntent(intent: Intent?) {
        val source = intent?.getStringExtra(EXTRA_SOURCE) ?: return
        // Consume so a config-change / re-delivery does not re-fire it.
        intent.removeExtra(EXTRA_SOURCE)
        val sink = notificationTapsSink
        if (sink != null) {
            sink.success(source)
        } else {
            pendingTapSource = source
        }
    }

    override fun onDetachedFromActivityForConfigChanges() = detachActivity()

    override fun onReattachedToActivityForConfigChanges(binding: ActivityPluginBinding) =
        onAttachedToActivity(binding)

    override fun onDetachedFromActivity() = detachActivity()

    private fun detachActivity() {
        activityBinding?.removeRequestPermissionsResultListener(this)
        activityBinding?.removeOnNewIntentListener(this)
        activity = null
        activityBinding = null
    }

    private fun streamHandler(onListen: (EventChannel.EventSink?) -> Unit) =
        object : EventChannel.StreamHandler {
            override fun onListen(arguments: Any?, events: EventChannel.EventSink?) {
                onListen(events)
            }

            override fun onCancel(arguments: Any?) {
                onListen(null)
            }
        }

    companion object {
        const val CHANNEL_METHODS = "makemind.appplayer_core/methods"
        const val CHANNEL_WAKES = "makemind.appplayer_core/wakes"
        const val CHANNEL_PERMISSION_CHANGES =
            "makemind.appplayer_core/permission_changes"
        const val CHANNEL_NOTIFICATION_TAPS =
            "makemind.appplayer_core/notification_taps"
        const val EXTRA_SOURCE = "com.makemind.appplayer_core.SOURCE"
    }
}

File: ios/Classes/AppPlayerCorePlugin.swift
import Flutter
import UIKit
import UserNotifications
import CoreLocation
import AVFoundation

/**
 * iOS side of the appplayer_core Platform Integration Foundation
 * (FR-PLATFORM): background execution (background task), OS permissions, and
 * notifications.
 */
public class AppPlayerCorePlugin: NSObject, FlutterPlugin,
    UNUserNotificationCenterDelegate {

  private var backgroundTaskId: UIBackgroundTaskIdentifier = .invalid
  private let notifications = NotificationBridge()

  private var wakesSink: FlutterEventSink?
  private var notificationTapsSink: FlutterEventSink?

  // A tap that arrived before Dart subscribed to the taps stream; flushed when
  // the sink attaches.
  private var pendingTapSource: String?

  public static func register(with registrar: FlutterPluginRegistrar) {
    let messenger = registrar.messenger()
    let instance = AppPlayerCorePlugin()

    let methodChannel = FlutterMethodChannel(
      name: "makemind.appplayer_core/methods", binaryMessenger: messenger)
    registrar.addMethodCallDelegate(instance, channel: methodChannel)

    FlutterEventChannel(name: "makemind.appplayer_core/wakes",
                        binaryMessenger: messenger)
      .setStreamHandler(WakesStreamHandler(instance))
    FlutterEventChannel(name: "makemind.appplayer_core/permission_changes",
                        binaryMessenger: messenger)
      .setStreamHandler(NoopStreamHandler())
    FlutterEventChannel(name: "makemind.appplayer_core/notification_taps",
                        binaryMessenger: messenger)
      .setStreamHandler(TapsStreamHandler(instance))

    // Receive notification taps so they can be routed back to the source app
    // (FR-NOTIF-004).
    UNUserNotificationCenter.current().delegate = instance
  }

  // MARK: - UNUserNotificationCenterDelegate

  public func userNotificationCenter(
    _ center: UNUserNotificationCenter,
    didReceive response: UNNotificationResponse,
    withCompletionHandler completionHandler: @escaping () -> Void
  ) {
    if let source =
      response.notification.request.content.userInfo["source"] as? String {
      if let sink = notificationTapsSink {
        sink(source)
      } else {
        pendingTapSource = source
      }
    }
    completionHandler()
  }

  public func userNotificationCenter(
    _ center: UNUserNotificationCenter,
    willPresent notification: UNNotification,
    withCompletionHandler completionHandler:
      @escaping (UNNotificationPresentationOptions) -> Void
  ) {
    // Show foreground notifications instead of suppressing them.
    if #available(iOS 14.0, *) {
      completionHandler([.banner, .sound])
    } else {
      completionHandler([.alert, .sound])
    }
  }

  public func handle(_ call: FlutterMethodCall, result: @escaping FlutterResult) {
    let args = call.arguments as? [String: Any] ?? [:]
    switch call.method {
    case "background.initialize":
      result(nil)
    case "background.begin":
      result(beginBackground())
    case "background.end":
      endBackground()
      result(nil)
    case "background.runJob":
      // iOS runs deferred work via BGTaskScheduler, registered by the host app
      // with its own identifiers. Acknowledge; the Dart scheduler reconciles on
      // the next foreground/wake pass.
      result(nil)
    case "permission.status":
      status(of: args["permission"] as? String, request: false, result: result)
    case "permission.request":
      status(of: args["permission"] as? String, request: true, result: result)
    case "notification.permissionStatus":
      notificationStatus(request: false, result: result)
    case "notification.requestPermission":
      notificationStatus(request: true, result: result)
    case "notification.post":
      notifications.post(
        id: args["id"] as? String ?? "",
        title: args["title"] as? String ?? "",
        body: args["body"] as? String ?? "",
        source: args["source"] as? String ?? "")
      result(nil)
    case "notification.cancel":
      notifications.cancel(id: args["id"] as? String ?? "")
      result(nil)
    default:
      result(FlutterMethodNotImplemented)
    }
  }

  // MARK: - Background execution

  private func beginBackground() -> Bool {
    endBackground()
    backgroundTaskId = UIApplication.shared.beginBackgroundTask(
      withName: "appplayer_core.background") { [weak self] in
      self?.endBackground()
    }
    return backgroundTaskId != .invalid
  }

  private func endBackground() {
    if backgroundTaskId != .invalid {
      UIApplication.shared.endBackgroundTask(backgroundTaskId)
      backgroundTaskId = .invalid
    }
  }

  // MARK: - Permissions

  private func status(of name: String?, request: Bool,
                      result: @escaping FlutterResult) {
    switch name {
    case "location":
      locationStatus(request: request, result: result)
    case "camera":
      mediaStatus(.video, request: request, result: result)
    case "microphone":
      mediaStatus(.audio, request: request, result: result)
    case "notifications":
      notificationStatus(request: request, result: result)
    default:
      // bluetooth (Info.plist backed), localNetwork, usb, backgroundExecution
      // have no simple gate on iOS — treated as granted.
      result("granted")
    }
  }

  private func locationStatus(request: Bool, result: @escaping FlutterResult) {
    let manager = CLLocationManager()
    let status: CLAuthorizationStatus
    if #available(iOS 14.0, *) {
      status = manager.authorizationStatus
    } else {
      status = CLLocationManager.authorizationStatus()
    }
    switch status {
    case .authorizedAlways, .authorizedWhenInUse:
      result("granted")
    case .denied:
      result("denied")
    case .restricted:
      result("restricted")
    case .notDetermined:
      if request { manager.requestWhenInUseAuthorization() }
      result("notDetermined")
    @unknown default:
      result("notDetermined")
    }
  }

  private func mediaStatus(_ type: AVMediaType, request: Bool,
                           result: @escaping FlutterResult) {
    let status = AVCaptureDevice.authorizationStatus(for: type)
    switch status {
    case .authorized: result("granted")
    case .denied: result("denied")
    case .restricted: result("restricted")
    case .notDetermined:
      if request {
        AVCaptureDevice.requestAccess(for: type) { granted in
          DispatchQueue.main.async { result(granted ? "granted" : "denied") }
        }
      } else {
        result("notDetermined")
      }
    @unknown default:
      result("notDetermined")
    }
  }

  private func notificationStatus(request: Bool, result: @escaping FlutterResult) {
    let center = UNUserNotificationCenter.current()
    if request {
      center.requestAuthorization(options: [.alert, .sound, .badge]) { granted, _ in
        DispatchQueue.main.async { result(granted ? "granted" : "denied") }
      }
      return
    }
    center.getNotificationSettings { settings in
      let value: String
      switch settings.authorizationStatus {
      case .authorized, .provisional, .ephemeral: value = "granted"
      case .denied: value = "denied"
      case .notDetermined: value = "notDetermined"
      @unknown default: value = "notDetermined"
      }
      DispatchQueue.main.async { result(value) }
    }
  }

  fileprivate func setWakesSink(_ sink: FlutterEventSink?) { wakesSink = sink }

  fileprivate func setTapsSink(_ sink: FlutterEventSink?) {
    notificationTapsSink = sink
    // Flush a tap captured before the stream was listened to.
    if let sink = sink, let pending = pendingTapSource {
      sink(pending)
      pendingTapSource = nil
    }
  }
}

// MARK: - Notification bridge

private class NotificationBridge {
  func post(id: String, title: String, body: String, source: String) {
    let content = UNMutableNotificationContent()
    content.title = title
    content.body = body
    content.userInfo = ["source": source]
    let request = UNNotificationRequest(
      identifier: id, content: content, trigger: nil)
    UNUserNotificationCenter.current().add(request)
  }

  func cancel(id: String) {
    UNUserNotificationCenter.current()
      .removePendingNotificationRequests(withIdentifiers: [id])
    UNUserNotificationCenter.current()
      .removeDeliveredNotifications(withIdentifiers: [id])
  }
}

// MARK: - Event stream handlers

private class WakesStreamHandler: NSObject, FlutterStreamHandler {
  private weak var plugin: AppPlayerCorePlugin?
  init(_ plugin: AppPlayerCorePlugin) { self.plugin = plugin }
  func onListen(withArguments arguments: Any?,
                eventSink events: @escaping FlutterEventSink) -> FlutterError? {
    plugin?.setWakesSink(events); return nil
  }
  func onCancel(withArguments arguments: Any?) -> FlutterError? {
    plugin?.setWakesSink(nil); return nil
  }
}

private class TapsStreamHandler: NSObject, FlutterStreamHandler {
  private weak var plugin: AppPlayerCorePlugin?
  init(_ plugin: AppPlayerCorePlugin) { self.plugin = plugin }
  func onListen(withArguments arguments: Any?,
                eventSink events: @escaping FlutterEventSink) -> FlutterError? {
    plugin?.setTapsSink(events); return nil
  }
  func onCancel(withArguments arguments: Any?) -> FlutterError? {
    plugin?.setTapsSink(nil); return nil
  }
}

private class NoopStreamHandler: NSObject, FlutterStreamHandler {
  func onListen(withArguments arguments: Any?,
                eventSink events: @escaping FlutterEventSink) -> FlutterError? { nil }
  func onCancel(withArguments arguments: Any?) -> FlutterError? { nil }
}
